### PR TITLE
Add `AsyncTemplateResponse` to Jinja2Templates

### PR DESCRIPTION
**Pull request by amisadmin**
Help Jinja2 make it easier and faster to use async template rendering.

### CODE REVIEW

**Review by yezz123 — APPROVED** [review]

File: tests/test_templates.py [hunk 1]
```diff
@@ -32,3 +32,34 @@ def test_template_response_requires_request(tmpdir):
...
     with pytest.raises(ValueError):
         templates.TemplateResponse("", {})
+
+
+def test_async_templates(tmpdir, test_client_factory):
+    path = os.path.join(tmpdir, "index.html")
+    with open(path, "w") as file:
+        file.write("<html>Hello,{{ get_name() }}</html>")
+
+    async def get_name():
```
Comment (aminalaee) [inline tests/test_templates.py:42]: Is this needed?

File: starlette/templating.py [hunk 2]
```diff
@@ -27,14 +27,14 @@ def __init__(
         self,
         template: typing.Any,
         context: dict,
+        content: typing.Optional[typing.Any] = None,
```
Comment (aminalaee) [inline starlette/templating.py]:
```suggestion
        content: typing.Optional[str] = None,
```

File: starlette/templating.py [hunk 3]
```diff
@@ -56,6 +56,12 @@ class Jinja2Templates:
     templates = Jinja2Templates("templates")
 
     return templates.TemplateResponse("index.html", {"request": request})
+
+    ..enable_async:
```
Comment (aminalaee) [inline starlette/templating.py:60]: Starlette is not using Sphinx, you can write a simple docstring and maybe add it to the `docs` directory with `Asynchronous template rendering`.

**Review by aminalaee — COMMENTED** [review]
I think this looks good, I'd wait for @tomchristie 's input on this too.

**Comment by euri10** [comment]
is this not valid anymore https://github.com/encode/starlette/blob/master/docs/templates.md#asynchronous-template-rendering ?

**Comment by aminalaee** [comment]
> https://github.com/encode/starlette/blob/master/docs/templates.md#asynchronous-template-rendering

Yeah I mentioned if this is going to be merged the docs should be updated.

File: starlette/templating.py (hunk 2 above)
Comment (amisadmin) [reply starlette/templating.py]: The `content` attribute in the parent `Response` is of type `Any`. I'm not sure it's necessarily of type `str` here.

File: tests/test_templates.py (hunk 1 above)
Comment (amisadmin) [reply tests/test_templates.py:42]: Yes.

File: starlette/templating.py (hunk 3 above)
Comment (amisadmin) [reply starlette/templating.py:60]: Ok, thanks for your advice. The document has been updated.

File: docs/templates.md [hunk 4]
```diff
@@ -88,4 +88,33 @@ Instead we'd recommend that you ensure that your endpoints perform all I/O,
...
+
+In asynchronous mode you can use 'AsyncTemplateResponse' to render asynchronous functions
+or coroutine objects in jinja2 templates.
+
+```python
+from starlette.templating import Jinja2Templates
+
+templates = Jinja2Templates(directory='templates', enable_async=True)
+
+async def get_name():
```
Comment (lovelydinosaur) [inline docs/templates.md:103]: I'd suggest some kind of comment noting that this'd actually only make sense in the case of an actual async operation. (Eg. an async database lookup.)

**Comment by lovelydinosaur** [comment]
> is this not valid anymore https://github.com/encode/starlette/blob/master/docs/templates.md#asynchronous-template-rendering ?

Subjective. Personally I think it's probably decent advice. But. 🤷‍♂️

File: starlette/templating.py (hunk 2 above)
Comment (aminalaee) [reply starlette/templating.py]: Well in this case it is always 'str' as output of 'render' sync and async in Jinja2.
I might be missing something but we could narrow the type down here as 'str' and that would still satisfy the parent with 'Any'.
What do you think @Kludex?

File: docs/templates.md (hunk 4 above)
Comment (amisadmin) [reply docs/templates.md:103]: Ok, thanks for your advice.

File: starlette/templating.py [hunk 5]
```diff
@@ -56,6 +56,12 @@ class Jinja2Templates:
...
+
+    templates = Jinja2Templates("templates",enable_async=True)
```
Comment (Kludex) [inline starlette/templating.py]:
```suggestion
    templates = Jinja2Templates("templates", enable_async=True)
```

File: starlette/templating.py [hunk 6]
```diff
@@ -95,9 +101,35 @@ def TemplateResponse(
...
+        name: str,
+        context: dict,
+        status_code: int = 200,
+        headers: typing.Optional[typing.Mapping[str, str]] = None,
+        media_type: typing.Optional[str] = None,
+        background: typing.Optional[BackgroundTask] = None,
+    ) -> _TemplateResponse:
+        if "request" not in context:
+            raise ValueError('context must include a "request" key')
+        assert self.env.is_async, "please enable async rendering mode."
```
Comment (Kludex) [inline starlette/templating.py]:
```suggestion
        if not self.env.is_async:
            raise RuntimeError("please enable async rendering mode.")
```

File: tests/test_templates.py [hunk 7]
```diff
@@ -32,3 +32,34 @@ def test_template_response_requires_request(tmpdir):
...
+
+    async def homepage(request):
+        return await templates.AsyncTemplateResponse("index.html", {"request": request})
+
+    app = Starlette(
+        debug=True,
+        routes=[Route("/", endpoint=homepage)],
+    )
+    templates = Jinja2Templates(directory=str(tmpdir), enable_async=True)
+    templates.env.globals["get_name"] = get_name
```
Comment (Kludex) [inline tests/test_templates.py:53]: Why is this not added on line 46?

**Review by Kludex — COMMENTED** [review]
We also need to tweak a bit the documentation, as the introduced section doesn't connect well with the previous section. 🤔

File: starlette/templating.py (hunk 2 above)
Comment (Kludex) [reply starlette/templating.py]: I'm not even sure if the `Optional` is needed. :sweat_smile: 

But yeah, it should be `str`. :+1: 

The return type of the render methods is `str`, and `_TemplateResponse` is a private class.

**Comment by Kludex** [comment]
@amisadmin Are you still interested on working on this PR?

**Comment by amisadmin** [comment]
> @amisadmin Are you still interested on working on this PR?

Yes, I am. But I'm a terrible document writer.

**Comment by Kludex** [comment]
I'm going to pass on this for the time being. We can come back to this after 1.0 is released.

Thanks for the PR, and sorry for the waiting time.

**Pull request closed without merging**